Issue-429 vSphere Validation Library and _test

tkrausjr opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide/pull-requests.md#the-pull-request-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
5. If this PR changes image versions, please title this PR "Bump <image name> from x.x.x to y.y.y."
-->

**What this PR does / why we need it**:
This PR adds a validation library in the util package that allows for pre validation of specific required objects in the VSphereMachineSpec and VSphereClusterSpec objects creating a MAP for each object with the individual status of the vsphere components in that object.  A validate_test.go file in included as well.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #429 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note

```

tkrausjr commented:
@akutz @andrewsykim  Can someone take a look at this.

akutz commented:
Hi @tkrausjr,

Yes, we will. Thank you for doing this. It’s only been a day since it was opened, so one of us will review it when we have a minute. I did take a quick look, and there are several comments I need to make.

Thank you again Thomas!

/assign

akutz commented:
/hold

akutz commented:
/ok-to-test

tkrausjr commented:
/retest

tkrausjr commented:
@akutz   Thanks I fixed the lint errors. Pipeline Job (pull-cluster-api-provider-vsphere-test ) is actually running the validate_test.go test which I wrote to actually test against a real environment. That environment does not exist in your test bed and I changed values to protect my environment so the test is failing.

tkrausjr commented:
/review

yastij commented:
/retest

yastij commented:
/test pull-cluster-api-provider-vsphere-e2e

yastij commented:
/retest

randomvariable commented on pkg/validate/validate.go:
```diff
@@ -0,0 +1,215 @@
+
+		return VSphereClusterStatus
+	}
+	// If you cant connect to VC then all VsphereCluster objects should return FailureMessage
```
Does this also occur when there's a transient network failure? Marking every machine in the cluster as failed as a result would be pretty destructive in that case.
